[Feat/schedule]  스케줄 기능 수정 사항 적용

BH13KDR opened the pull request:
# ✅ test: Django shell환경에서 Test완료 했습니다.

## ✅ PR 요약
- 작업 요약: 완성 된 schedule 생성, 조회, 상세조회, 수정, 삭제 기능을 완성, 테스트 했습니다.

## 📄 상세 내용
- [x] 주요 변경 사항 1 : django shell환경에서 테스트가 완수되었습니다.
- [x] 주요 변경 사항 2 : 각종 N+1문제를 select_related, prefetch_related 등을 적용해 해결했습니다.
- [x] 주요 변경 사항 3 : 각각 전달하던 데이터를 request.data로 일괄 전달하게 바꾸었습니다.
- [x] 주요 변경 사항 4 : 중복되던 코드를 리팩토링하고, 불필요한 코드를 제거했습니다.
- [x] 주요 변경 사항 5 : 유효성 검증 방식을 통일 했습니다.
- [x] 주요 변경 사항 6 : 에러 케이스 테스트 코드도 추가하였습니다.

Commit:
# ✨ feat: StudyGroup : Schedule기능을 완성했습니다.
# ✅ test: Django shell환경에서 Test완료 했습니다.

SWWS97 commented on apps/study_groups/serializers/schedule_serializers.py:
```diff
@@ -0,0 +1,109 @@
+    # session_date 검증
+    def validate_session_date(self, value: Union[str, datetime]) -> datetime:
+        # 문자열로 들어온 경우 처리
+        if isinstance(value, str):
```
위에서 이미 session_date = serializers.DateTimeField() 이렇게 선언하셔서
여기 검증은 필요없을거 같아요 형식 체크는 빼주세용

SWWS97 commented on apps/study_groups/services/schedule_services.py:
```diff
@@ -0,0 +1,77 @@
+    # 스케줄 조회
+    @staticmethod
+    def list_schedules(*, group_id: int) -> List[GroupSchedule]:
+        return list(GroupSchedule.objects.filter(study_group_id=group_id))
```
여기 prefetch_related 걸어주세요

SWWS97 commented on apps/study_groups/services/schedule_services.py:
```diff
@@ -0,0 +1,77 @@
+    @staticmethod
+    def retrieve_schedule(*, schedule_id: int) -> GroupSchedule:
+        try:
+            return GroupSchedule.objects.get(id=schedule_id)
```
여기도 마찬가지로 select_related, prefetch_related 걸어서 N+1 해결해주세요

SWWS97 commented on apps/study_groups/serializers/schedule_serializers.py:
```diff
@@ -0,0 +1,109 @@
+
+from apps.study_groups.models import (
+    GroupMember,
+    GroupSchedule,
```
안쓰는 import는 지워주세요

SWWS97 commented on apps/study_groups/views/schedule_views.py:
```diff
@@ -0,0 +1,185 @@
+        study_group = StudyGroup.objects.get(id=group_id)
+
+        serializer = GroupScheduleSerializer(
+            data={
```
이거 이렇게 안해도 data=request.data로 사용할 수 있어요

SWWS97 commented on apps/study_groups/views/schedule_views.py:
```diff
@@ -0,0 +1,185 @@
+
+        serializer = GroupScheduleSerializer(
+            schedule,
+            data=data,
```
여기도 위에서 data로 재구성 할 필요없이 data부분 지우고 data=request.data 이렇게 써주세요

SWWS97 commented on apps/study_groups/views/schedule_views.py:
```diff
@@ -0,0 +1,185 @@
+        )
+        serializer.is_valid(raise_exception=True)
+
+        user_id = cast(int, request.user.id)
```
여기 user_id는 안쓰이는거 같아요

wjdtjddns98 commented on apps/study_groups/serializers/schedule_serializers.py:
```diff
@@ -0,0 +1,109 @@
+                raise serializers.ValidationError({"participants": f"유효하지 않은 스터디 그룹 멤버 id: {ids_str}"})
+
+        return attrs
+
```
참가자 검증 로직 N+1 문제 해결 해서 다시 작성 해주시면 좋을 것 같습니다
지금은 괜찮지만 현재 로직은 대규모 데이터 일때는 쿼리에 부하를 줄 수 있는 로직입니당

SWWS97 reviewed (changes requested):
현재 테스트에 성공일 때 케이스만 있어서 에러 케이스일 때 테스트도 추가해주세요

wjdtjddns98 commented on apps/study_groups/services/schedule_services.py:
```diff
@@ -0,0 +1,77 @@
+
+    @staticmethod
+    def delete_schedule(*, schedule: GroupSchedule) -> None:
+        schedule.delete()
```
create와 update 메서드 간의 코드 중복을 제거해서 DRY 원칙을 준수해서 리팩토링 해보시는게 어떨까용

wjdtjddns98 commented on apps/study_groups/views/schedule_views.py:
```diff
@@ -0,0 +1,185 @@
+            context={"study_group": study_group},
+        )
+        serializer.is_valid(raise_exception=True)
+
```
시리얼라이저에 데이터 전달할 때 request.data의 필드를 일일이 추출하여 새로운 딕셔너리를 만들 필요 없어요. DRF의 표준 패턴은 request.data를 바로 시리얼라이저에 전달하는데 이렇게 하면 코드가 더 간결해지고 유지보수성이 높아짐니다

data = request.data.copy()
serializer = GroupScheduleSerializer(
    data=data,
    context={"study_group": study_group},
)

이렇게 맨 마지막에 context만 하나 추가해주면 좋을 것 같아요

wjdtjddns98 replied:
put 메서드에도 마찬가지로 똑같이 해주세용

wjdtjddns98 commented on apps/study_groups/views/schedule_views.py:
```diff
@@ -0,0 +1,185 @@
+            {"data": GroupScheduleSerializer(schedule).data},
+            status=status.HTTP_201_CREATED,
+        )
+
```
왜 이쪽 유효성 검증만 커스텀 안하신건지 따로 이유가 있나요 ?

Commit:
# ✅ test     : 성공 테스트 외, 실패시 테스트도 추가.
# ♻️ refactor : 리팩터링 : 각종 N+1 문제를 해결했습니다. 불필요한 코드를 제거했습니다.request.data로 data를 일괄 전달하게 바꾸었습니다. 중복되던 코드를 리팩토링했습니다. 유효성 검증 방식을 통일했습니다.

Commit:
CI TEST

Commit:
CI test2

BH13KDR commented:
다시 PR하겠습니다.
